### DISCUSSION TITLE
react-spring simple examples

### DISCUSSION THREAD

**Discussion by bradwoods** [pmndrs/react-spring#972] [Show and tell]
I started a list of simple examples of react-spring for anyone getting started. Let me know if you are after anything specific and I'll add it.

https://examples.bradwoods.io/react-spring

**Comment by aomini** [comment]
hey I tried to copy your first and second example to learn but I'm getting error on `wxyz.to()` function.
can you look at this sandbox once [bardwoods demo copy](https://codesandbox.io/s/react-spring-gz4nm?file=/src/App.js:452-557)

**Reply by bradwoods** [reply]
hey mate, that error is occurring because of the react-spring version being used. Could you upgrade it to 9.0.0-beta.34 & see if that clears it up.

**Reply by aomini** [reply]
ok will do that. @bradwoods. Hey brad have you used it with styled-components? I dont seem to find a proper example of styled-components and react-spring as well as can we use the react-spring in hover  property of css?

**Reply by bradwoods** [reply]
I've added an example using styled-components. Let me know if it is what you were after. Could you give more details about the hover you mentioned?

**Reply by aomini** [reply]
Oh that's great. Thank you.
About hover I meant can we use or do something like:
```js
const AnimatedDiv = styled(animated.div)`
    background: tomato;
    &:hover{
     ${props => props.customStyles}
   }
`
const spring = useSpring({opacity: 1, from :{opacity: 1}})
//use it like
<AnimatedDiv customStyles={spring}/>
```

**Reply by bradwoods** [reply]
Would the end result be to trigger a spring-animation when hovered?

**Reply by aomini** [reply]
yes. I was wondering if we can do that without react events.

**Reply by bradwoods** [reply]
I'm not sure if that is possible. This is how I would approach it - https://codesandbox.io/s/with-styled-component-hover-f4gg6?file=/src/index.tsx

**Reply by aomini** [reply]
coool. thanks please keep updating your site. bookmarked it.

**Reply by bradwoods** [reply]
Thanks for the feedback mate, will do.

**Reply by aomini** [reply]
hey can you add a example of `useTrails` with alternate and repeating animation demo? I don't know how to implement that.

**Reply by bradwoods** [reply]
Sure. I've added it under _transition-trail-repeating_. Let me know if that is what you were after.

**Reply by aomini** [reply]
oh yeah thanks. So `onRest` method runs every time whenever the animation ends. But isnt updating stating like will rerenderthe entire component? Can we just rerender the animated div only?
Thanks alot for your examples.

**Reply by bradwoods** [reply]
No worries. I'm glad someone is finding them useful! To my knowledge, I don't think it's possible to rerun the animation without rerendering the component using react-spring. The animation needs a value to change in order to restart. When that value changes, the component re-renders.

**Reply by aomini** [reply]
that's cool.. but react-spring have keyframes... can you do that with keyframes as well?? the docs seems to be more confusing with keyframes.

**Reply by bradwoods** [reply]
Here's an [example](https://bradwoods.io/guides/react-spring/keyframes-animation) that uses keyframes with onRest to reset, is that what your looking for?

**Comment by AndrewSB** [comment]
I'd love to see an example with an externally defined svg file (see https://github.com/react-spring/react-spring/discussions/989)

**Reply by bradwoods** [reply]
I'm not sure how you would go about that. In my experience, when animating svgs, I have always had them inline so I can access the different elements that make up the image. I've added an example of animating an inline svg stroke in case you decide to use that approach.

**Comment by bradwoods** [comment]
Added an example of animating a number. https://bradwoods.io/guides/react-spring/animate-a-number

**Reply by aomini** [reply]
great one. Thanks. I learnt a lot from your examples. 
hey Bradwood, I found this template on gatsby which was created with react-spring. Can you have examples with this sort of parallax effect? https://rakeshshubhu.github.io/
keep on uploading examples man.

**Reply by bradwoods** [reply]
Glad to hear it. Here a parallax example. I hope it's useful.  https://bradwoods.io/guides/react-spring/parallax-scroll.

**Reply by aomini** [reply]
wow.. That's so cool. thanks.

**Reply by aomini** [reply]
why did you use  useCallback here? 
```js
const onScroll = useCallback(
		e => set({ scrollTop: e.target.scrollTop }),
		[set]
	);
```

**Reply by bradwoods** [reply]
Just as an optimization. The docs talk about it [here](https://reactjs.org/docs/hooks-reference.html#usecallback).

**Comment by bradwoods** [comment]
Added a config example - https://bradwoods.io/guides/react-spring/config

**Comment by bradwoods** [comment]
Added a config presets example - https://bradwoods.io/guides/react-spring/config-presets

**Comment by sanderwapstra** [comment]
@bradwoods Thanks for these examples, really helpful! I'm struggling with a repeating useChain. I want to animate three dots (scale to 1.2 and back) and the second dot is starting before the first one is finished so it's a nice wave. Do you know / have an example what's best practice for these kind of animations? :)

**Reply by bradwoods** [reply]
Hey @sanderwapstra. Glad there useful! Are you looking to animate the dots so the 2nd dots starts its animation when the 1st dot's animation finishes?

**Reply by sanderwapstra** [reply]
@bradwoods Thanks for your reply! No, the second one starts before the first one is finished. So there are three elements animating, but I’m not sure how to loop a group of animations.

**Reply by bradwoods** [reply]
Does this example help? https://bradwoods.io/guides/react-spring/transition-trailing-repeating
It sounds like trail would be a good fit for this scenario.

**Reply by sanderwapstra** [reply]
With useTrail I cannot set a custom delay I believe (https://github.com/react-spring/react-spring/discussions/1083#discussioncomment-39267), but your suggestion (the `onRest` function) combined with useTransition works! Thanks!

**Comment by phoenisx** [comment]
Hi @bradwoods 

Thanks for your examples.

I am struggling to create an animated Window Scrollspy. Sandbox here - https://codesandbox.io/s/react-spring-scrollspy-gei79

The issue is, if you click on `Page 3` link, it will animate properly to `.page-3`.

The issue starts, when user manually scrolls to some other place, say, the user clicks on `Page 3` link and manually scrolls back to website top: 0, and clicks `Page 1`. Clicking on `Page 1` should have animated from top to bottom, but it animates from last scroll position, i.e. it animates from bottom to top, `Page 3` to `Page 1`.

Not sure, what I am doing wrong, or what particular API or config I am missing, due to which this is troubling me.

**Reply by phoenisx** [reply]
Ok no worries!! The moment I was writing on this thread, I got a solution.


I just had to use `set({ from: { y: window.pageYOffset }, to: { y } });`, to specify the initial Y offset, from which to animate.

I wasted a whole day for such a simple answer. 😓

**Reply by bradwoods** [reply]
Glad you worked it out, looked tricky.

**Comment by bradwoods** [comment]
Added a chaining transitions example - https://bradwoods.io/guides/react-spring/transition-chaining

**Reply by hedberg** [reply]
Hi @bradwoods,
thanks for sharing your work! 
I'm looking to create a (out-in) transition between a number of screens/elements so that the active one fades out before the next one animates in. Seems like an easy enough thing to do but I haven't been able to work it out so far..  Do you have an idea of how that should be solved?

**Comment by tcherokee** [comment]
Hi Brad, 

I am hoping you can help me with an example. I am trying to replicate [this animation](https://codesandbox.io/s/morning-snow-e26b9?file=/index.html) which I did using CSS Transitions in React Spring. So far I have gotten this far [https://codesandbox.io/s/suspicious-tharp-yru1v?file=/src/App.js](https://codesandbox.io/s/suspicious-tharp-yru1v?file=/src/App.js). I can technically finish the animation by using two more useSpring functions to animate the bottom span, but considering the bottom span is using the same animation as the top span, but just with negative values instead of positive values, I figure there should be a more efficient way of doing this?

**Reply by bradwoods** [reply]
hey mate,

I would use the same approach as you. I think useChain is the best fit for this problem. I don't see any way of making this more efficient without it becoming needlessly complicated. What I would do from here is make 1 react component for the bottom line & 1 for the top line. Then create the springs within each of those components.

**Reply by tcherokee** [reply]
Thanks Brad, it is good to see I was on the right track :). Finished up the animation. you can see it here [https://codesandbox.io/s/animated-hamburger-menu-yru1v?file=/src/App.js](https://codesandbox.io/s/animated-hamburger-menu-yru1v?file=/src/App.js). Maybe someone will find it useful

**Reply by tcherokee** [reply]
Actually, I was playing around a bit more with react spring and I think I might have found a way to shorten the about of code I needed for this animation. If I use the shorthand style props, I can reverse the animation by adding `-` to the variable in the style section. so something like this seems to work ```transform: xyz.to((x, y, z) => `translate3d(${x}px, ${-y}px, ${z}px)`)```. However, the issue I am running into now is using two transforms in the style tag, it seems react spring might be ignoring the first transform and only running the second one. You can see an example in my playground here [https://codesandbox.io/s/peaceful-meadow-4og6b?file=/src/App.js:1188-1258](https://codesandbox.io/s/peaceful-meadow-4og6b?file=/src/App.js:1188-1258)

**Reply by bradwoods** [reply]
Yeah the 2nd transform would override the 1st one.

**Reply by tcherokee** [reply]
Thanks for clarifying, is there any way to combine transforms?

**Reply by bradwoods** [reply]
Not to my knowledge, considering each transform is taking values from separate functions

**Reply by tcherokee** [reply]
Ok, thank you for all your help

**Comment by bradwoods** [comment]
New examples design out. Let me know if I can change anything to make things easier. https://bradwoods.io/guides/react-spring

**Comment by bradwoods** [comment]
Added how to reset a transition example - https://bradwoods.io/guides/react-spring/transition-reset

**Comment by bradwoods** [comment]
Added on-frame callback example - https://bradwoods.io/guides/react-spring/callback-on_frame

**Reply by ramojej** [reply]
hi @bradwoods,

just want to ask what's the best practice for animating on scroll? Is using react-intersection-observer in combination with react-spring a good practice? Thanks!

**Reply by bradwoods** [reply]
Hey mate,

The only work I've done in regards to scroll is this [parallax scroll example](https://bradwoods.io/guides/react-spring/parallax-scroll). I think it would depend on what the goal is. Could you give some details?

**Reply by ramojej** [reply]
Well, I would just like to animate elements when the user scrolls to a certain section. For example I scroll down to "Services section" maybe with an offset of 100px, only then the elements would animate.. each with a different delay.

**Reply by bradwoods** [reply]
I think the intersection observer might be your best bet for that.

**Reply by ramojej** [reply]
Do you have like a really simple slide up effect example?

**Reply by bradwoods** [reply]
I'm not sure what you mean, could you give more detail?

**Comment by bradwoods** [comment]
Added animation on-complete callback example - https://bradwoods.io/guides/react-spring/callback-on_complete

**Comment by bradwoods** [comment]
Using react-spring in react-three-fiber - https://bradwoods.io/guides/react-spring/with-react-three-fiber

**Comment by joakim-roos** [comment]
Thanks for these examples, getting into react-spring now!
Do you have any tips on how to do a simplified version of this interpolation animation without using use-tweak? Onscroll-event or a looping animation would be really cool.  https://codesandbox.io/s/use-tweaks-example-ekqv2

**Reply by bradwoods** [reply]
No worries. I'm not sure what your after. Are you looking to scroll the text left / right on a loop, without using the control panel?

**Comment by bradwoods** [comment]
Use duration - https://examples.bradwoods.io/react-spring/transition-duration

**Comment by stressed-pet-owner** [comment]
Hi, Brad!

Would like to know how I can make this kind of animation possible.. I only animated simple ones and being challenged by this one: https://share.getcloudapp.com/X6uexJr6

Will it be better to animate it as a single SVG or separate SVGs?

**Reply by bradwoods** [reply]
Hey mate, 

I think trailing would be a good fit for this one. Here's an [example](https://examples.bradwoods.io/react-spring/transition-trailing). I think I would use a single SVG. It might make things easier when dealing with the lines that lead to the center.

**Reply by stressed-pet-owner** [reply]
Thank you for your input.. Is it possible with that kind of popup effect? I just don't know how can I create it with the "whole" svg... do you have other example? Here's the SVG by the way :))

https://drive.google.com/file/d/1D8rQpDV0EKgYnekkV9rw5Mb-l6IYzrNb/view?usp=sharing

**Reply by bradwoods** [reply]
I think it would be possible. The way I would do it is by applying springs to elements within the SVG, like `circle` or a group nesting multiple elements. Sorry I don't have any examples for that specific case.

**Reply by stressed-pet-owner** [reply]
@bradwoods,

tried my best, but can't copy the effect in the video.. how to do the "popup effect". what I did was scale it from 0 to 1.. but it's sliding. The lines also I can't animate..

My biggest issues it to make it automatically animate without pressing any button. Can you give me some tips please. Thanks a lot!

here's the sandbox: https://codesandbox.io/s/peaceful-drake-zgyuh?file=/src/AnimatedSVG.js

**Reply by bradwoods** [reply]
To make an animation start automatically you can use the `from` property, [example](https://examples.bradwoods.io/react-spring/transition-on-mount).

**Reply by stressed-pet-owner** [reply]
Thanks. :)

**Reply by stressed-pet-owner** [reply]
Hi, mate @bradwoods..

Can you please help me with the draw line effect? I've tried, but can't make it work..
